### PR TITLE
More configurable collecting metadata for backup

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Improvement

### Changelog entry:
More configurable collecting metadata for backup.

Now there will be the following settings in the configuration file:

```
<backups>
    <collect_metadata_timeout>600000</collect_metadata_timeout>
    <attempts_to_collect_metadata_before_sleep>2</attempts_to_collect_metadata_before_sleep>
    <min_sleep_before_next_attempt_to_collect_metadata>100</min_sleep_before_next_attempt_to_collect_metadata>
    <max_sleep_before_next_attempt_to_collect_metadata>5000</max_sleep_before_next_attempt_to_collect_metadata>
    <compare_collected_metadata>1</compare_collected_metadata>
</backups>
```

Here
- `collect_metadata_timeout` is the time the BACKUP command will try to collect the metadata of tables & databases;
- `attempts_to_collect_metadata_before_sleep` is the number of attempts to collect the metadata before sleeping;
- `min_sleep_before_next_attempt_to_collect_metadata` is the minimum time clickhouse will wait after unsuccessful attempt before trying to collect the metadata again;
- `max_sleep_before_next_attempt_to_collect_metadata` is the maximum time clickhouse will wait after unsuccessful attempt before trying to collect the metadata again;
- `compare_collected_metadata` is whether clickhouse should collect the metadata after a successful attempt one more time and check that nothing has changed. `compare_collected_metadata = 1` makes that metadata collecting more reliable, so don't set `compare_collected_metadata` unless you're sure you need it. (Because if you set `compare_collected_metadata` to `0` then some recently added tables can be not included in the backup so it's a dangerous option.)

All times are in milliseconds.

With the default values the metadata collecting works in the following way:
1) it tries to collect the metadata for the first time; then
2) it tries to collect it again, and compares the results from the first and the second collecting; if they match, it's done; otherwise
3) it sleeps 100 millisecond and tries to collect again, then compares the results from the second and the third collecting; if they match, it's done; otherwise
4) it tries to collect again, then compares the results from the third and the fourth collecting; if they match, it's done; otherwise
5) it sleeps 200 milliseconds and tries to collect again, then compares the results from the fourh and the fifth collecting; if they match, it's done;
...
and so on, the sleep time is doubled each time until it reaches 5000 milliseconds.
And such attempts will be continued until 600000 milliseconds pass.

For most users the default values must be good enough.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit fb1c7ba294222701c2153ce3191ca406fbe7c4de with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54804/fb1c7ba294222701c2153ce3191ca406fbe7c4de/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: src/Backups/BackupEntriesCollector.cpp [hunk 1]
```diff
@@ -251,12 +250,72 @@ bool BackupEntriesCollector::tryGatherMetadataAndCompareWithPrevious(std::option
             throw;
 
         inconsistency_error = e;
-        return false;
+        need_another_attempt = true;
```
Comment (SmitaRKulkarni) [inline src/Backups/BackupEntriesCollector.cpp:253]: If `compare_collected_metadata` is not true, then we don't need another attempt right ?

File: src/Backups/BackupEntriesCollector.h [hunk 2]
```diff
@@ -97,8 +98,24 @@ class BackupEntriesCollector : private boost::noncopyable
     std::shared_ptr<IBackupCoordination> backup_coordination;
     const ReadSettings read_settings;
     ContextPtr context;
-    std::chrono::milliseconds on_cluster_first_sync_timeout;
-    std::chrono::milliseconds consistent_metadata_snapshot_timeout;
+
+    const std::chrono::milliseconds on_cluster_first_sync_timeout;
```
Comment (SmitaRKulkarni) [inline src/Backups/BackupEntriesCollector.h:102]: For my understanding, why do we need different timeout for first cluster sync ?

File: src/Backups/BackupEntriesCollector.cpp (hunk 1 above)
Comment (vitlibar) [reply src/Backups/BackupEntriesCollector.cpp:253]: The functions `gatherTablesMetadata()` and `gatherDatabasesMetadata()` can throw the `INCONSISTENT_METADATA_FOR_BACKUP`, see for example https://github.com/ClickHouse/ClickHouse/blob/af8b1e4978519b23a8c550bd3df25a354ed799ca/src/Databases/DatabasesCommon.cpp#L352. In that case the gathered data are not consistent by itself and we need another attempt even if `compare_collected_metadata` is true.

File: src/Backups/BackupEntriesCollector.h (hunk 2 above)
Comment (vitlibar) [reply src/Backups/BackupEntriesCollector.h:102]: The first cluster sync is a bit different. It means the time required to send the query to the cluster's nodes plus the time for the nodes to start working on the backup. By default `on_cluster_first_sync_timeout` is 180 milliseconds which corresponds to [distributed_ddl_task_timeout](https://github.com/ClickHouse/ClickHouse/blob/af8b1e4978519b23a8c550bd3df25a354ed799ca/src/Core/Settings.h#L305).

File: src/Backups/BackupEntriesCollector.cpp (hunk 1 above)
Comment (vitlibar) [reply src/Backups/BackupEntriesCollector.cpp:255]: I've added this comment to the code.

File: src/Backups/BackupEntriesCollector.h (hunk 2 above)
Comment (vitlibar) [reply src/Backups/BackupEntriesCollector.h:104]: I've added this comment to the code.

**Review by SmitaRKulkarni — APPROVED** [review]
LGTM

**Pull request merged**